sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/VariantExtractionImpl.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.read;
+
+import java.util.Arrays;
+import java.util.Objects;
+
+import org.apache.spark.annotation.Experimental;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.Metadata;
+
+/**
+ * Default implementation of {@link VariantExtraction}.
+ *
+ * @since 4.1.0
+ */
+@Experimental
+public final class VariantExtractionImpl implements VariantExtraction {
```

Comment:
I don't see the value of separating the interface and imp if we need to expose the impl anyway. Can we follow `TableInfo` and make `VariantExtraction` a class directly?

Reply:
but let me confirm again: the API means Spark passes `VariantExtraction` instances to the data source, and data sources need to return bool. Why do data sources need to use the impl class?